crates/nu-command/src/filters/items.rs

```diff
@@ -0,0 +1,163 @@
+        let capture_block: Closure = call.req(engine_state, stack, 0)?;
+
+        let metadata = input.metadata();
+        let ctrlc = engine_state.ctrlc.clone();
+        let engine_state = engine_state.clone();
+        let block = engine_state.get_block(capture_block.block_id).clone();
+        let mut stack = stack.captures_to_stack(&capture_block.captures);
+        let orig_env_vars = stack.env_vars.clone();
+        let orig_env_hidden = stack.env_hidden.clone();
+        let span = call.head;
+        let redirect_stdout = call.redirect_stdout;
+        let redirect_stderr = call.redirect_stderr;
+
+        let input_span = input.span().unwrap_or(Span::unknown());
+        let run_for_each_item = move |keyval: (String, Value)| -> Option<Value> {
+            // with_env() is used here to ensure that each iteration uses
+            // a different set of environment variables.
+            // Hence, a 'cd' in the first loop won't affect the next loop.
+            stack.with_env(&orig_env_vars, &orig_env_hidden);
+
+            if let Some(var) = block.signature.get_positional(0) {
+                if let Some(var_id) = &var.var_id {
+                    stack.add_var(*var_id, Value::string(keyval.0.clone(), span));
+                }
+            }
+
+            if let Some(var) = block.signature.get_positional(1) {
+                if let Some(var_id) = &var.var_id {
+                    stack.add_var(*var_id, keyval.1);
+                }
+            }
+
+            match eval_block_with_early_return(
+                &engine_state,
+                &mut stack,
+                &block,
+                PipelineData::empty(),
+                redirect_stdout,
+                redirect_stderr,
+            ) {
+                Ok(v) => Some(v.into_value(span)),
+                Err(ShellError::Break(_)) => None,
+                Err(error) => {
+                    let error = chain_error_with_input(error, Ok(input_span));
+                    Some(Value::Error {
+                        error: Box::new(error),
+                    })
+                }
+            }
+        };
+        match input {
+            PipelineData::Empty => Ok(PipelineData::Empty),
+            PipelineData::Value(Value::Record { cols, vals, .. }, ..) => Ok(cols
+                .into_iter()
+                .zip(vals.into_iter())
+                .into_iter()
+                .map_while(run_for_each_item)
+                .into_pipeline_data(ctrlc)),
+            PipelineData::ListStream(stream, ..) => {
+                let v: Vec<_> = stream.into_iter().collect();
+                let cols = get_columns(&v);
+                let vals = get_values(&v, call.head, input_span)?;
+
+                Ok(cols
+                    .into_iter()
+                    .zip(vals.into_iter())
+                    .into_iter()
+                    .map_while(run_for_each_item)
+                    .into_pipeline_data(ctrlc))
+            }
```

Comment:
I think this would be inconsistent behavior. This accepts a table if the list is a `PipelineData::ListStream` but not if it is a `...::Value(Value::List{})` of records.

Reply:
Based on the type error messages I would say lets remove this match arm and only handle records and `LazyRecord`s right now.